### ISSUE TITLE
Internal error with `data .. where` definitions

### ISSUE THREAD

**Issue by iitalics** [agda/agda#3892]
The following code causes an internal error:

```agda
open import Data.Nat.Base using (ℕ)

Foo : ℕ
Foo = 5 -- the body of this definition doesn't matter

data Foo where
{- An internal error has occurred. Please report this as a bug.
Location of the error: src/full/Agda/TypeChecking/Rules/Data.hs:75 -}
```

This should instead trigger an error explaining that `Foo` was not previously declared as a data type. Note that we do get the correct error message if we don't give Foo a body:

```agda
Foo : ℕ

data Foo where
{- Foo has been declared as a function, but is being defined as a data
type -}
```

```
$ agda -V
Agda version 2.6.1-27a1a2a
```

**Comment by andreasabel** [comment]
MWE:
```agda
postulate Foo : Set
data Foo where
```
This is a regression in 2.5.4.  Version 2.5.3 reports:
```
Missing type signature for Foo
```

**Comment by andreasabel** [comment]
A related problem is
```agda
data D : Set where
  foo : D

data D where
```
which only complains about the duplicate _module_:
```
Duplicate definition of module D. Previous definition of (datatype)
module D at ...
when scope checking the declaration
  data D where
```
This suboptimal error message was actually introduced in #2576.

**Comment by andreasabel** [comment]
I am heaping more heuristics onto the scope checker to fix this error (#2576 really).  But the root cause is the ugly nicifier (Concrete.Definitions).
